Fix wasm compilation and send body and headers from WasmClient

Perseus101 opened the pull request:
Fix wasm compilation

* Upgrade to http-types v2.3.0 to get the wasm compilation fix
* Add CI job to check compilation on wasm32-unknown-unknown
* Closes #3

FreddieRidell committed:
Send body and headers correctly in the wasm backend

Perseus101 committed:
Fix compilation on wasm32-unknown-unknown

Perseus101 committed:
Add CI job for wasm32-unknown-unknown

Perseus101 committed:
Update sending body and headers in the wasm backend

Perseus101 committed:
Bump http-types to the latest version

yoshuawuyts commented on Cargo.toml:
```diff
@@ -42,6 +42,7 @@ isahc = { version = "0.9", optional = true, default-features = false, features =
 js-sys = { version = "0.3.25", optional = true }
 wasm-bindgen = { version = "0.2.48", optional = true }
 wasm-bindgen-futures = { version = "0.4.5", optional = true }
+futures-util = "0.3.1"
```
I'd rather we not introduce `futures-util` into this crate; it comes with a lot of extra dependencies that have a negative impact on compilation times.

yoshuawuyts replied:
If we have to, I'd rather we rely on `async-std` with `default-features = false` and `features = ["core"]` enabled. That will keep our dependency tree to a minimum in most cases.

Perseus101 replied:
With some minor tweaks, I've removed futures-util.

yoshuawuyts commented on src/wasm.rs:
```diff
@@ -33,17 +34,19 @@ impl HttpClient for WasmClient {
+                let name: http_types::headers::HeaderName = name.parse().unwrap();
+                response.insert_header(
+                    &name,
+                    value.parse::<http_types::headers::HeaderValue>().unwrap(),
```
Casting `value` to a `String` or `&str` is enough to pass it to `insert_header`. `value.parse()` shouldn't be needed here.

Perseus101 replied:
Good catch, thanks!

yoshuawuyts commented on src/wasm.rs:
```diff
@@ -74,44 +76,85 @@ impl Future for InnerFuture {
-        init: RequestInit,
-        url: String,
+        request: web_sys::Request,
+        _body_buf: Pin<Vec<u8>>,
```
This name should not have a leading `_` since we're using it.
```suggestion
        body_buf: Pin<Vec<u8>>,
```

Perseus101 replied:
This strictly isn't being used in Rust, it's only being held to ensure the buffer isn't deallocated while the request is completed in Javascript. However, I also don't like the leading `_`, so I changed the name and added a `#[allow(dead_code)]` to suppress the warning.

yoshuawuyts commented on src/wasm.rs:
```diff
@@ -74,44 +76,85 @@ impl Future for InnerFuture {
+    use std::pin::Pin;
 
     /// Create a new fetch request.
-    pub(crate) fn new(method: impl AsRef<str>, url: impl AsRef<str>) -> Request {
-        Request::new(method, url)
-    }
```
I like this better indeed; good call.

Perseus101 replied:
Full credit to @FreddieRidell :+1:

yoshuawuyts commented on src/wasm.rs:
```diff
@@ -74,44 +76,85 @@ impl Future for InnerFuture {
+            let uri = req.url().to_string();
+            let mut body = req.take_body();
+
+            //convert the body into a uint8 array
```
```suggestion
            // convert the body into a uint8 array
```

yoshuawuyts commented on src/wasm.rs:
```diff
@@ -74,44 +76,85 @@ impl Future for InnerFuture {
+
+            Ok(Self {
+                request,
+                _body_buf: body_pinned,
```
```suggestion
                body_buf: body_pinned,
```

yoshuawuyts commented on src/wasm.rs:
```diff
@@ -74,44 +76,85 @@ impl Future for InnerFuture {
+            let mut body = req.take_body();
+
+            //convert the body into a uint8 array
+            // needs to be pinned and retained inside the Request because the Uint8Array passed to
+            // js is just a portal into WASM linear memory, and if the underlying data is moved the
+            // js ref will become silently invalid
+            let mut body_buf = Vec::with_capacity(1024);
```
Clever; I wasn't aware of this. Is this what was causing the bug?

Perseus101 replied:
This is part of the fix for #3, originally from PR #20. I forgot to update the title and description when I rebased #20, so I didn't give you all the context for this fix. Sorry about that!

yoshuawuyts reviewed (approved):
Thanks so much for this PR. This looks great overall! -- CI is currencly failing, and there are a few nits -- but otherwise this is :100:!

Perseus101 committed:
Remove futures-util and apply notes from code review

Co-authored-by: Yoshua Wuyts <yoshuawuyts+github@gmail.com>

Perseus101 commented:
I've rebased and updated the changes from #20 into this PR. Those changes needed a little updating, since the library has changed a bit since that PR was created, but the majority of the code still worked and I was able to update it without too much effort.

#20 can be closed after this is merged.

Perseus101 commented:
@yoshuawuyts I've fixed the nits you pointed out.

~~The CI failure had nothing to do with this pull request. It originates in `cookie` v0.14.1 (https://github.com/SergioBenitez/cookie-rs/issues/153), which was pulled in when `tide` was updated to 0.9 in dev-dependencies. Running `cargo +stable doc` on master fails for the same reason, so I'm assuming this made it to master by accident. Unfortunately, the latest version of `http-types` also depends on this version of `cookie`, so I can't just downgrade `tide` to fix the error.~~

~~To circumvent the error, I changed the CI to run `cargo doc --no-deps` instead of `cargo doc` so that only the docs for this crate are built. The other option would be to run `cargo doc` on nightly, which is what `http-types` does.~~

~~I wanted to draw attention to this because changing the CI so that the tests pass feels a little like cheating :sweat_smile:, but I feel it's justified in this case.~~

EDIT: CI was fixed in #31

Perseus101 committed:
Merge branch 'master' into fix-wasm-compilation

Perseus101 committed:
Bump http-types to 2.3.0

Perseus101 committed:
Update errors to http_types errors

Perseus101 commented:
@yoshuawuyts 
Bumping this since I haven't heard back in a while. This should be ready to merge.

yoshuawuyts reviewed (approved):
This looks great; thanks heaps!

yoshuawuyts commented:
@Fishrock123 I think this patch may be a good candidate to cut a new release for; wdyt?

Fishrock123 commented:
If this confirms that WASM does indeed compile via CI, then I think that would be a reasonable idea.

Perseus101 commented:
I'd like to echo the request for a new release. I have a PR open on surf waiting on a new release of http-client to get these fixes merged: https://github.com/http-rs/surf/pull/185
